sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -267,12 +267,32 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   }
 
   object CartesianProduct extends Strategy {
+    def createCartesianProduct(left: LogicalPlan, right: LogicalPlan): SparkPlan = {
+      // If plan can broadcast we use BroadcastNestedLoopJoin, as we know for inner join with true
+      // condition is same as Cartesian.
+      // For CartesianProduct we will use the small size plan as cartesian left rdd.
+      if (right.statistics.sizeInBytes <= left.statistics.sizeInBytes) {
+        right match {
+          case CanBroadcast(right) => joins.BroadcastNestedLoopJoin(planLater(left),
+              planLater(right), joins.BuildRight, Inner, None)
+          case _ => execution.joins.CartesianProduct(planLater(left), planLater(right),
+            joins.BuildRight)
+        }
+      } else {
+        left match {
+          case CanBroadcast(left) => joins.BroadcastNestedLoopJoin(planLater(left),
+            planLater(right), joins.BuildLeft, Inner, None)
```

Comment:
I think `BroadcastNestedLoopJoin` can support the condition, and push down the filter into the operator can also reduce the memory overhead, as `BroadcastNestedLoopJoin` will put all of the valid tuple into a compact buffer.